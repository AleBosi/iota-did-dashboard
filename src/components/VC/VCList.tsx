import { useState } from "react";
import { VerifiableCredential } from "../../models/vc";
import VCViewer from "./VCViewer";

interface Props {
  vcs: VerifiableCredential[];
  onSelect?: (vc: VerifiableCredential) => void;
}

export default function VCList({ vcs, onSelect }: Props) {
  const [selected, setSelected] = useState<VerifiableCredential | null>(null);

  return (
    <div className="flex gap-8">
      <ul className="w-1/2">
        {vcs.map(vc => (
          <li
            key={vc.id}
            className="mb-2 border-b pb-2 cursor-pointer hover:bg-blue-50"
            onClick={() => {
              setSelected(vc);
              onSelect?.(vc);
            }}
          >
            <span className="font-semibold">{vc.type[1] || vc.type[0]}</span>
            <span className="ml-2 text-gray-400">{vc.issuer}</span>
            <span className="ml-2 text-xs text-gray-400">{vc.issuanceDate}</span>
            {/* Status estratto da credentialSubject */}
            {vc.credentialSubject?.status && (
              <span className={`ml-2 text-xs ${vc.credentialSubject.status === "valid" ? "text-green-600" : "text-red-600"}`}>
                {vc.credentialSubject.status}
              </span>
            )}
          </li>
        ))}
      </ul>
      <div className="w-1/2">
        {/* Mappatura dati per VCViewer */}
        {selected && (
          <VCViewer
            vc={{
              subject: selected.credentialSubject.id,
              type: selected.type[1] || selected.type[0],
              value: selected.credentialSubject.value,
              issuedAt: selected.issuanceDate,
              issuer: selected.issuer
            }}
          />
        )}
      </div>
    </div>
  );
}
