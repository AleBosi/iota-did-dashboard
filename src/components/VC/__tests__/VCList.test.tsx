import { render, screen, fireEvent } from "@testing-library/react";
import VCList from "../VCList";

// Mock completo VerifiableCredential (con status, value nel credentialSubject)
const mockVCs = [
  {
    '@context': ["https://www.w3.org/2018/credentials/v1"],
    id: "1",
    type: ["VerifiableCredential", "Compliance"],
    issuer: "did:iota:issuer1",
    issuanceDate: "2025-08-01",
    credentialSubject: {
      id: "did:iota:prod1",
      value: "ok",
      status: "valid"
    },
    proof: {
      type: "Ed25519Signature2020",
      created: "2025-08-01T00:00:00Z",
      proofPurpose: "assertionMethod",
      verificationMethod: "did:iota:issuer1#keys-1",
      jws: "mock",
      hash: "mock-hash-1"
    }
  },
  {
    '@context': ["https://www.w3.org/2018/credentials/v1"],
    id: "2",
    type: ["VerifiableCredential", "Quality"],
    issuer: "did:iota:issuer2",
    issuanceDate: "2025-08-02",
    credentialSubject: {
      id: "did:iota:prod2",
      value: "not ok",
      status: "revoked"
    },
    proof: {
      type: "Ed25519Signature2020",
      created: "2025-08-02T00:00:00Z",
      proofPurpose: "assertionMethod",
      verificationMethod: "did:iota:issuer2#keys-1",
      jws: "mock",
      hash: "mock-hash-2"
    }
  }
];

// Mock del componente VCViewer per isolare il test
jest.mock("../VCViewer", () => ({
  __esModule: true,
  default: ({ vc }: any) => <div data-testid="vcviewer">{vc?.subject}</div>
}));

describe("VCList component", () => {
  it("renderizza tutte le VC ricevute", () => {
    render(<VCList vcs={mockVCs} />);
    expect(screen.getByText("Compliance")).toBeInTheDocument();
    expect(screen.getByText("Quality")).toBeInTheDocument();
    expect(screen.getByText("did:iota:issuer1")).toBeInTheDocument();
    expect(screen.getByText("did:iota:issuer2")).toBeInTheDocument();
    expect(screen.getByText("2025-08-01")).toBeInTheDocument();
    expect(screen.getByText("2025-08-02")).toBeInTheDocument();
  });

  it("status ha classe corretta: verde per valid, rossa per revoked", () => {
    render(<VCList vcs={mockVCs} />);
    const statusValid = screen.getByText("valid");
    const statusRevoked = screen.getByText("revoked");
    expect(statusValid).toHaveClass("text-green-600");
    expect(statusRevoked).toHaveClass("text-red-600");
  });

  it("se clicco una VC, mostra il dettaglio in VCViewer", () => {
    render(<VCList vcs={mockVCs} />);
    fireEvent.click(screen.getByText("Compliance"));
    expect(screen.getByTestId("vcviewer")).toHaveTextContent("did:iota:prod1");
    fireEvent.click(screen.getByText("Quality"));
    expect(screen.getByTestId("vcviewer")).toHaveTextContent("did:iota:prod2");
  });

  it("chiama onSelect quando una VC viene cliccata", () => {
    const onSelect = jest.fn();
    render(<VCList vcs={mockVCs} onSelect={onSelect} />);
    fireEvent.click(screen.getByText("Compliance"));
    expect(onSelect).toHaveBeenCalledWith(expect.objectContaining({ id: "1" }));
    fireEvent.click(screen.getByText("Quality"));
    expect(onSelect).toHaveBeenCalledWith(expect.objectContaining({ id: "2" }));
  });

  it("non esplode se la lista Ã¨ vuota", () => {
    render(<VCList vcs={[]} />);
    expect(screen.queryByText("Compliance")).toBeNull();
    expect(screen.queryByText("Quality")).toBeNull();
  });

  it("matcha lo snapshot", () => {
    const { asFragment } = render(<VCList vcs={mockVCs} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
