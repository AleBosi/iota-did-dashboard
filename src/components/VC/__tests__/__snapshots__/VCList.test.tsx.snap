// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`VCList component matcha lo snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex gap-8"
  >
    <ul
      class="w-1/2"
    >
      <li
        class="mb-2 border-b pb-2 cursor-pointer hover:bg-blue-50"
      >
        <span
          class="font-semibold"
        >
          Compliance
        </span>
        <span
          class="ml-2 text-gray-400"
        >
          did:iota:issuer1
        </span>
        <span
          class="ml-2 text-xs text-gray-400"
        >
          2025-08-01
        </span>
        <span
          class="ml-2 text-xs text-green-600"
        >
          valid
        </span>
      </li>
      <li
        class="mb-2 border-b pb-2 cursor-pointer hover:bg-blue-50"
      >
        <span
          class="font-semibold"
        >
          Quality
        </span>
        <span
          class="ml-2 text-gray-400"
        >
          did:iota:issuer2
        </span>
        <span
          class="ml-2 text-xs text-gray-400"
        >
          2025-08-02
        </span>
        <span
          class="ml-2 text-xs text-red-600"
        >
          revoked
        </span>
      </li>
    </ul>
    <div
      class="w-1/2"
    />
  </div>
</DocumentFragment>
`;
